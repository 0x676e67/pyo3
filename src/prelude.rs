// Copyright (c) 2017-present PyO3 Project and Contributors

//! A collection of items you most likely want to have in scope when working with pyo3
//!
//! The purpose of this module is to alleviate imports of many common pyo3 traits
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use conversion::{FromPyObject, IntoPyObject, PyTryFrom, PyTryInto, ToPyObject};
pub use err::{PyErr, PyResult};
pub use instance::{AsPyRef, Py, PyToken};
pub use noargs::NoArgs;
pub use object::PyObject;
pub use objectprotocol::ObjectProtocol;
pub use python::Python;
pub use pythonrun::GILGuard;
// This is only part of the prelude because we need it for the pymodinit function
pub use types::PyModule;
// This is required for the constructor
pub use PyRawObject;

pub use pyo3cls::{pyclass, pyfunction, pymethods, pyproto};

#[cfg(Py_3)]
pub use pyo3cls::mod3init as pymodinit;

#[cfg(not(Py_3))]
pub use pyo3cls::mod2init as pymodinit;
