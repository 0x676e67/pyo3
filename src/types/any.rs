use crate::conversion::PyTryFrom;
use crate::err::PyDowncastError;
use crate::internal_tricks::Unsendable;
use crate::{ffi, PyObject};

/// Represents a python's [Any](https://docs.python.org/3/library/typing.html#typing.Any) type.
/// We can convert all python objects as `PyAny`.
///
/// In addition, if the inner object is an instance of type `T`, we can downcast
/// `PyAny` into `T`.
///
/// # Example
///
/// ```
/// use pyo3::prelude::*;
/// use pyo3::types::{PyAny, PyDict, PyList};
/// let gil = Python::acquire_gil();
/// let dict = PyDict::new(gil.python());
/// assert!(gil.python().is_instance::<PyAny, _>(dict).unwrap());
/// let any = dict.as_ref();
/// assert!(any.downcast_ref::<PyDict>().is_ok());
/// assert!(any.downcast_ref::<PyList>().is_err());
/// ```
#[repr(transparent)]
pub struct PyAny(PyObject, Unsendable);
unsafe impl crate::type_object::PyObjectLayout<PyAny> for ffi::PyObject {
    unsafe fn unchecked_ref(&self) -> &PyAny {
        &*((&self) as *const &Self as *const _)
    }
    unsafe fn unchecked_refmut(&mut self) -> &mut PyAny {
        &mut *((&self) as *const &mut Self as *const _ as *mut _)
    }
}
impl crate::type_object::PyObjectSizedLayout<PyAny> for ffi::PyObject {}
pyobject_native_type_named!(PyAny);
pyobject_native_type_convert!(
    PyAny,
    ffi::PyObject,
    ffi::PyBaseObject_Type,
    Some("builtins"),
    ffi::PyObject_Check
);
pyobject_native_type_extract!(PyAny);

impl PyAny {
    pub fn downcast<T>(&self) -> Result<&T, PyDowncastError>
    where
        for<'py> T: PyTryFrom<'py>,
    {
        <T as PyTryFrom>::try_from(self)
    }
}
