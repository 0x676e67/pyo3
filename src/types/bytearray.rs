// Copyright (c) 2017-present PyO3 Project and Contributors
use crate::err::{PyErr, PyResult};
use crate::instance::PyNativeType;
use crate::{ffi, AsPyPointer, PyAny, Python};
use std::os::raw::c_char;
use std::slice;

/// Represents a Python `bytearray`.
#[repr(transparent)]
pub struct PyByteArray(PyAny);

pyobject_native_var_type!(PyByteArray, ffi::PyByteArray_Type, ffi::PyByteArray_Check);

impl PyByteArray {
    /// Creates a new Python bytearray object.
    ///
    /// The byte string is initialized by copying the data from the `&[u8]`.
    pub fn new<'p>(py: Python<'p>, src: &[u8]) -> &'p PyByteArray {
        let ptr = src.as_ptr() as *const c_char;
        let len = src.len() as ffi::Py_ssize_t;
        unsafe { py.from_owned_ptr::<PyByteArray>(ffi::PyByteArray_FromStringAndSize(ptr, len)) }
    }

    /// Creates a new Python `bytearray` object with an `init` closure to write its contents.
    /// Before calling `init` the bytearray is zero-initialised.
    /// * If `init` returns `Err(e)`, `new_with` will return `Err(e)`.
    /// * If `init` returns `Ok(new_len)`, the allocated bytearray will be truncated to length
    ///   `new_len` and `new_with` will return `Ok(&PyByteArray)`.
    /// * If `init` returns `Ok(None)`, `new_with` will return `Ok(&PyByteArray)`.
    ///
    /// Iff Python raises a MemoryError on the allocation, `new_with` will return it inside `Err`.
    ///
    /// # Example
    /// ```
    /// use pyo3::{prelude::*, types::PyByteArray};
    /// Python::with_gil(|py| -> PyResult<()> {
    ///     let py_bytearray = PyByteArray::new_with(py, 10, |bytes: &mut [u8]| {
    ///         bytes.copy_from_slice(b"Hello Rust");
    ///         Ok(None)
    ///     })?;
    ///     let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
    ///     assert_eq!(bytearray, b"Hello Rust");
    ///     Ok(())
    /// });
    /// ```
    pub fn new_with<F>(py: Python, len: usize, init: F) -> PyResult<&PyByteArray>
    where
        F: FnOnce(&mut [u8]) -> PyResult<Option<usize>>,
    {
        unsafe {
            let pyptr =
                ffi::PyByteArray_FromStringAndSize(std::ptr::null(), len as ffi::Py_ssize_t);
            if pyptr.is_null() {
                return Err(PyErr::fetch(py));
            }
            let buffer = ffi::PyByteArray_AsString(pyptr) as *mut u8;
            debug_assert!(!buffer.is_null());
            // Zero-initialise the uninitialised bytearray
            std::ptr::write_bytes(buffer, 0u8, len);
            // (Further) Initialise the bytearray in init
            match init(std::slice::from_raw_parts_mut(buffer, len)) {
                Ok(Some(new_len)) if new_len < len => {
                    if ffi::PyByteArray_Resize(pyptr, new_len as ffi::Py_ssize_t) != 0 {
                        return Err(PyErr::fetch(py));
                    }
                }
                Ok(_) => (),
                Err(e) => {
                    // Deallocate pyptr to avoid leaking the bytearray
                    ffi::Py_DECREF(pyptr);
                    return Err(e);
                }
            };
            Ok(py.from_owned_ptr(pyptr))
        }
    }

    /// Creates a new Python bytearray object from another PyObject that
    /// implements the buffer protocol.
    pub fn from<'p, I>(py: Python<'p>, src: &'p I) -> PyResult<&'p PyByteArray>
    where
        I: AsPyPointer,
    {
        unsafe { py.from_owned_ptr_or_err(ffi::PyByteArray_FromObject(src.as_ptr())) }
    }

    /// Gets the length of the bytearray.
    #[inline]
    pub fn len(&self) -> usize {
        // non-negative Py_ssize_t should always fit into Rust usize
        unsafe { ffi::PyByteArray_Size(self.as_ptr()) as usize }
    }

    /// Checks if the bytearray is empty.
    pub fn is_empty(&self) -> bool {
        self.len() == 0
    }

    /// Get the start of the buffer containing the contents of the bytearray.
    ///
    /// Note that this bytearray object is both shared and mutable, and the backing buffer may be
    /// reallocated if the bytearray is resized. This can occur from Python code as well as from
    /// Rust via [PyByteArray::resize].
    ///
    /// As a result, the returned pointer should be dereferenced only if since calling this method
    /// no Python code has executed, [PyByteArray::resize] has not been called.
    pub fn data(&self) -> *mut u8 {
        unsafe { ffi::PyByteArray_AsString(self.as_ptr()) as *mut u8 }
    }

    /// Get the contents of this buffer as a slice.
    ///
    /// # Safety
    /// This bytearray must not be resized or edited while holding the slice.
    ///
    /// ## Safety Detail
    /// This method is equivalent to `std::slice::from_raw_parts(self.data(), self.len())`, and so
    /// all the safety notes of `std::slice::from_raw_parts` apply here.
    ///
    /// In particular, note that this bytearray object is both shared and mutable, and the backing
    /// buffer may be reallocated if the bytearray is resized. Mutations can occur from Python
    /// code as well as from Rust, via [PyByteArray::as_bytes_mut] and [PyByteArray::resize].
    ///
    /// Extreme care should be exercised when using this slice, as the Rust compiler will
    /// make optimizations based on the assumption the contents of this slice cannot change. This
    /// can easily lead to undefined behavior.
    ///
    /// As a result, this slice should only be used for short-lived operations to read this
    /// bytearray without executing any Python code, such as copying into a Vec.
    pub unsafe fn as_bytes(&self) -> &[u8] {
        slice::from_raw_parts(self.data(), self.len())
    }

    /// Get the contents of this buffer as a mutable slice.
    ///
    /// # Safety
    /// This slice should only be used for short-lived operations that write to this bytearray
    /// without executing any Python code. See the safety note for [PyByteArray::as_bytes].
    #[allow(clippy::mut_from_ref)]
    pub unsafe fn as_bytes_mut(&self) -> &mut [u8] {
        slice::from_raw_parts_mut(self.data(), self.len())
    }

    /// Copies the contents of the bytearray to a Rust vector.
    ///
    /// # Example
    ///
    /// ```
    /// # use pyo3::prelude::*;
    /// # use pyo3::types::PyByteArray;
    /// # use pyo3::types::IntoPyDict;
    /// # let gil = GILGuard::acquire();
    /// # let py = gil.python();
    /// #
    /// let bytearray = PyByteArray::new(py, b"Hello World.");
    /// let mut copied_message = bytearray.to_vec();
    /// assert_eq!(b"Hello World.", copied_message.as_slice());
    ///
    /// copied_message[11] = b'!';
    /// assert_eq!(b"Hello World!", copied_message.as_slice());
    ///
    /// let locals = [("bytearray", bytearray)].into_py_dict(py);
    /// py.run("assert bytearray == b'Hello World.'", None, Some(locals)).unwrap();
    /// ```
    pub fn to_vec(&self) -> Vec<u8> {
        unsafe { self.as_bytes() }.to_vec()
    }

    /// Resizes the bytearray object to the new length `len`.
    ///
    /// Note that this will invalidate any pointers obtained by [PyByteArray::data], as well as
    /// any (unsafe) slices obtained from [PyByteArray::as_bytes] and [PyByteArray::as_bytes_mut].
    pub fn resize(&self, len: usize) -> PyResult<()> {
        unsafe {
            let result = ffi::PyByteArray_Resize(self.as_ptr(), len as ffi::Py_ssize_t);
            if result == 0 {
                Ok(())
            } else {
                Err(PyErr::fetch(self.py()))
            }
        }
    }
}

#[cfg(test)]
mod test {
    use crate::exceptions;
    use crate::object::PyObject;
    use crate::types::PyByteArray;
    use crate::Python;

    #[test]
    fn test_len() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        let src = b"Hello Python";
        let bytearray = PyByteArray::new(py, src);
        assert_eq!(src.len(), bytearray.len());
    }

    #[test]
    fn test_as_bytes() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        let src = b"Hello Python";
        let bytearray = PyByteArray::new(py, src);

        let slice = unsafe { bytearray.as_bytes() };
        assert_eq!(src, slice);
        assert_eq!(bytearray.data() as *const _, slice.as_ptr());
    }

    #[test]
    fn test_as_bytes_mut() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        let src = b"Hello Python";
        let bytearray = PyByteArray::new(py, src);

        let slice = unsafe { bytearray.as_bytes_mut() };
        assert_eq!(src, slice);
        assert_eq!(bytearray.data(), slice.as_mut_ptr());

        slice[0..5].copy_from_slice(b"Hi...");

        assert_eq!(
            bytearray.str().unwrap().to_str().unwrap(),
            "bytearray(b'Hi... Python')"
        );
    }

    #[test]
    fn test_to_vec() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        let src = b"Hello Python";
        let bytearray = PyByteArray::new(py, src);

        let vec = bytearray.to_vec();
        assert_eq!(src, vec.as_slice());
    }

    #[test]
    fn test_from() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        let src = b"Hello Python";
        let bytearray = PyByteArray::new(py, src);

        let ba: PyObject = bytearray.into();
        let bytearray = PyByteArray::from(py, &ba).unwrap();

        assert_eq!(src, unsafe { bytearray.as_bytes() });
    }

    #[test]
    fn test_from_err() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        if let Err(err) = PyByteArray::from(py, &py.None()) {
            assert!(err.is_instance::<exceptions::PyTypeError>(py));
        } else {
            panic!("error");
        }
    }

    #[test]
    fn test_resize() {
        let gil = Python::acquire_gil();
        let py = gil.python();

        let src = b"Hello Python";
        let bytearray = PyByteArray::new(py, src);

        bytearray.resize(20).unwrap();
        assert_eq!(20, bytearray.len());
    }

    #[test]
    fn test_byte_array_new_with() -> super::PyResult<()> {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let py_bytearray = PyByteArray::new_with(py, 10, |b: &mut [u8]| {
            b.copy_from_slice(b"Hello Rust");
            Ok(None)
        })?;
        let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
        assert_eq!(bytearray, b"Hello Rust");
        Ok(())
    }

    #[test]
    fn test_byte_array_new_with_zero_initialised() -> super::PyResult<()> {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let py_bytearray = PyByteArray::new_with(py, 10, |_b: &mut [u8]| Ok(None))?;
        let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
        assert_eq!(bytearray, &[0; 10]);
        Ok(())
    }

    #[test]
    fn test_byte_array_new_with_truncated() -> super::PyResult<()> {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let py_bytearray = PyByteArray::new_with(py, 10, |b: &mut [u8]| {
            b.copy_from_slice(b"Hello Rust");
            Ok(Some(5))
        })?;
        let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
        assert_eq!(bytearray, b"Hello");
        Ok(())
    }

    #[test]
    fn test_byte_array_new_with_no_growing() -> super::PyResult<()> {
        let gil = Python::acquire_gil();
        let py = gil.python();
        let py_bytearray = PyByteArray::new_with(py, 10, |b: &mut [u8]| {
            b.copy_from_slice(b"Hello Rust");
            Ok(Some(15))
        })?;
        let bytearray: &[u8] = unsafe { py_bytearray.as_bytes() };
        assert_eq!(bytearray, b"Hello Rust");
        Ok(())
    }

    #[test]
    fn test_byte_array_new_with_error() -> super::PyResult<()> {
        use crate::exceptions::PyValueError;
        use crate::types::PyDict;

        let alloc_size = 1024 * 1024;

        let gil = Python::acquire_gil();
        let py = gil.python();
        let locals = PyDict::new(py);

        py.run(
            r#"
import tracemalloc

tracemalloc.start()
memory_usage_before = sum(
    alloc.size
    for alloc in tracemalloc.take_snapshot()
    .filter_traces((tracemalloc.Filter(True, "<unknown>"),))
    .statistics("filename")
)
"#,
            None,
            Some(locals),
        )?;

        let err = PyValueError::py_err("Hello Crustaceans!");
        let py_bytearray_result = PyByteArray::new_with(py, alloc_size, |_b: &mut [u8]| Err(err));

        assert!(py_bytearray_result.is_err());
        assert!(py_bytearray_result
            .err()
            .unwrap()
            .is_instance::<PyValueError>(py));

        py.run(
            r#"
memory_usage_after = sum(
    alloc.size
    for alloc in tracemalloc.take_snapshot()
    .filter_traces((tracemalloc.Filter(True, "<unknown>"),))
    .statistics("filename")
)
tracemalloc.stop()
"#,
            None,
            Some(locals),
        )?;

        let memory_usage_before: usize =
            locals.get_item("memory_usage_before").unwrap().extract()?;
        let memory_usage_after: usize = locals.get_item("memory_usage_after").unwrap().extract()?;

        assert!((memory_usage_after - memory_usage_before) < alloc_size);

        Ok(())
    }
}
