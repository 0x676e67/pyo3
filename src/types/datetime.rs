use err::PyResult;
use ffi;
use ffi::PyDateTimeAPI;
use ffi::{PyDateTime_Check, PyDateTime_DateTimeType};
use ffi::{
    PyDateTime_DATE_GET_HOUR, PyDateTime_DATE_GET_MICROSECOND, PyDateTime_DATE_GET_MINUTE,
    PyDateTime_DATE_GET_SECOND,
};
use ffi::{
    PyDateTime_DELTA_GET_DAYS, PyDateTime_DELTA_GET_MICROSECONDS, PyDateTime_DELTA_GET_SECONDS,
};
use ffi::{PyDateTime_DateType, PyDate_Check};
use ffi::{PyDateTime_DeltaType, PyDelta_Check};
use ffi::{PyDateTime_GET_DAY, PyDateTime_GET_MONTH, PyDateTime_GET_YEAR};
use ffi::{
    PyDateTime_TIME_GET_HOUR, PyDateTime_TIME_GET_MICROSECOND, PyDateTime_TIME_GET_MINUTE,
    PyDateTime_TIME_GET_SECOND,
};
use ffi::{PyDateTime_TZInfoType, PyTZInfo_Check};
use ffi::{PyDateTime_TimeType, PyTime_Check};
use object::PyObject;
use std::os::raw::c_int;

#[cfg(Py_3_6)]
use ffi::{PyDateTime_DATE_GET_FOLD, PyDateTime_TIME_GET_FOLD};

use instance::Py;
use python::{Python, ToPyPointer};

// Traits
pub trait PyDateAccess {
    fn get_year(&self) -> i32;
    fn get_month(&self) -> u8;
    fn get_day(&self) -> u8;
}

pub trait PyDeltaAccess {
    fn get_days(&self) -> i32;
    fn get_seconds(&self) -> i32;
    fn get_microseconds(&self) -> i32;
}

pub trait PyTimeAccess {
    fn get_hour(&self) -> u8;
    fn get_minute(&self) -> u8;
    fn get_second(&self) -> u8;
    fn get_microsecond(&self) -> u32;
    #[cfg(Py_3_6)]
    fn get_fold(&self) -> u8;
}

// datetime.date bindings
pub struct PyDate(PyObject);
pyobject_native_type!(PyDate, PyDateTime_DateType, PyDate_Check);

impl PyDate {
    pub fn new(py: Python, year: i32, month: u8, day: u8) -> PyResult<Py<PyDate>> {
        unsafe {
            let ptr = (PyDateTimeAPI.Date_FromDate)(
                year,
                c_int::from(month),
                c_int::from(day),
                PyDateTimeAPI.DateType,
            );
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }

    pub fn from_timestamp(py: Python, args: &PyObject) -> PyResult<Py<PyDate>> {
        unsafe {
            let ptr = (PyDateTimeAPI.Date_FromTimestamp)(PyDateTimeAPI.DateType, args.as_ptr());
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }
}

impl PyDateAccess for PyDate {
    fn get_year(&self) -> i32 {
        unsafe { PyDateTime_GET_YEAR(self.as_ptr()) as i32 }
    }

    fn get_month(&self) -> u8 {
        unsafe { PyDateTime_GET_MONTH(self.as_ptr()) as u8 }
    }

    fn get_day(&self) -> u8 {
        unsafe { PyDateTime_GET_DAY(self.as_ptr()) as u8 }
    }
}

// datetime.datetime bindings
pub struct PyDateTime(PyObject);
pyobject_native_type!(PyDateTime, PyDateTime_DateTimeType, PyDateTime_Check);

impl PyDateTime {
    pub fn new(
        py: Python,
        year: i32,
        month: u8,
        day: u8,
        hour: u8,
        minute: u8,
        second: u8,
        microsecond: u32,
        tzinfo: Option<&PyObject>,
    ) -> PyResult<Py<PyDateTime>> {
        unsafe {
            let ptr = (PyDateTimeAPI.DateTime_FromDateAndTime)(
                year,
                c_int::from(month),
                c_int::from(day),
                c_int::from(hour),
                c_int::from(minute),
                c_int::from(second),
                microsecond as c_int,
                opt_to_pyobj(py, tzinfo),
                PyDateTimeAPI.DateTimeType,
            );
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }

    pub fn from_timestamp(
        py: Python,
        args: &PyObject,
        kwargs: &PyObject,
    ) -> PyResult<Py<PyDateTime>> {
        unsafe {
            let ptr = (PyDateTimeAPI.DateTime_FromTimestamp)(
                PyDateTimeAPI.DateTimeType,
                args.as_ptr(),
                kwargs.as_ptr(),
            );
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }
}

impl PyDateAccess for PyDateTime {
    fn get_year(&self) -> i32 {
        unsafe { PyDateTime_GET_YEAR(self.as_ptr()) as i32 }
    }

    fn get_month(&self) -> u8 {
        unsafe { PyDateTime_GET_MONTH(self.as_ptr()) as u8 }
    }

    fn get_day(&self) -> u8 {
        unsafe { PyDateTime_GET_DAY(self.as_ptr()) as u8 }
    }
}

impl PyTimeAccess for PyDateTime {
    fn get_hour(&self) -> u8 {
        unsafe { PyDateTime_DATE_GET_HOUR(self.as_ptr()) as u8 }
    }

    fn get_minute(&self) -> u8 {
        unsafe { PyDateTime_DATE_GET_MINUTE(self.as_ptr()) as u8 }
    }

    fn get_second(&self) -> u8 {
        unsafe { PyDateTime_DATE_GET_SECOND(self.as_ptr()) as u8 }
    }

    fn get_microsecond(&self) -> u32 {
        unsafe { PyDateTime_DATE_GET_MICROSECOND(self.as_ptr()) as u32 }
    }

    #[cfg(Py_3_6)]
    fn get_fold(&self) -> u8 {
        unsafe { PyDateTime_DATE_GET_FOLD(self.as_ptr()) as u8 }
    }
}

// datetime.time
pub struct PyTime(PyObject);
pyobject_native_type!(PyTime, PyDateTime_TimeType, PyTime_Check);

impl PyTime {
    pub fn new(
        py: Python,
        hour: u8,
        minute: u8,
        second: u8,
        microsecond: u32,
        tzinfo: Option<&PyObject>,
    ) -> PyResult<Py<PyTime>> {
        unsafe {
            let ptr = (PyDateTimeAPI.Time_FromTime)(
                c_int::from(hour),
                c_int::from(minute),
                c_int::from(second),
                microsecond as c_int,
                opt_to_pyobj(py, tzinfo),
                PyDateTimeAPI.TimeType,
            );
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }

    #[cfg(Py_3_6)]
    pub fn new_with_fold(
        py: Python,
        hour: u8,
        minute: u8,
        second: u8,
        microsecond: u32,
        tzinfo: Option<&PyObject>,
        fold: bool,
    ) -> PyResult<Py<PyTime>> {
        unsafe {
            let ptr = (PyDateTimeAPI.Time_FromTimeAndFold)(
                c_int::from(hour),
                c_int::from(minute),
                c_int::from(second),
                microsecond as c_int,
                opt_to_pyobj(py, tzinfo),
                fold as c_int,
                PyDateTimeAPI.TimeType,
            );
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }
}

impl PyTimeAccess for PyTime {
    fn get_hour(&self) -> u8 {
        unsafe { PyDateTime_TIME_GET_HOUR(self.as_ptr()) as u8 }
    }

    fn get_minute(&self) -> u8 {
        unsafe { PyDateTime_TIME_GET_MINUTE(self.as_ptr()) as u8 }
    }

    fn get_second(&self) -> u8 {
        unsafe { PyDateTime_TIME_GET_SECOND(self.as_ptr()) as u8 }
    }

    fn get_microsecond(&self) -> u32 {
        unsafe { PyDateTime_TIME_GET_MICROSECOND(self.as_ptr()) as u32 }
    }

    #[cfg(Py_3_6)]
    fn get_fold(&self) -> u8 {
        unsafe { PyDateTime_TIME_GET_FOLD(self.as_ptr()) as u8 }
    }
}

// datetime.tzinfo bindings
pub struct PyTzInfo(PyObject);
pyobject_native_type!(PyTzInfo, PyDateTime_TZInfoType, PyTZInfo_Check);

// datetime.timedelta bindings
pub struct PyDelta(PyObject);
pyobject_native_type!(PyDelta, PyDateTime_DeltaType, PyDelta_Check);

impl PyDelta {
    pub fn new(
        py: Python,
        days: i32,
        seconds: i32,
        microseconds: i32,
        normalize: bool,
    ) -> PyResult<Py<PyDelta>> {
        unsafe {
            let ptr = (PyDateTimeAPI.Delta_FromDelta)(
                days as c_int,
                seconds as c_int,
                microseconds as c_int,
                normalize as c_int,
                PyDateTimeAPI.DeltaType,
            );
            Py::from_owned_ptr_or_err(py, ptr)
        }
    }
}

impl PyDeltaAccess for PyDelta {
    fn get_days(&self) -> i32 {
        unsafe { PyDateTime_DELTA_GET_DAYS(self.as_ptr()) as i32 }
    }

    fn get_seconds(&self) -> i32 {
        unsafe { PyDateTime_DELTA_GET_SECONDS(self.as_ptr()) as i32 }
    }

    fn get_microseconds(&self) -> i32 {
        unsafe { PyDateTime_DELTA_GET_MICROSECONDS(self.as_ptr()) as i32 }
    }
}

// Utility function
unsafe fn opt_to_pyobj(py: Python, opt: Option<&PyObject>) -> *mut ffi::PyObject {
    // Convenience function for unpacking Options to either an Object or None
    match opt {
        Some(tzi) => tzi.as_ptr(),
        None => py.None().as_ptr(),
    }
}
