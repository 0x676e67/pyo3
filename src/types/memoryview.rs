use crate::err::PyResult;
use crate::{ffi, AsPyPointer, PyAny};

/// Represents a Python `memoryview`.
#[repr(transparent)]
pub struct PyMemoryView(PyAny);

pyobject_native_type_core!(PyMemoryView, pyobject_native_static_type_object!(ffi::PyMemoryView_Type), #checkfunction=ffi::PyMemoryView_Check);

impl PyMemoryView {
    /// Creates a new Python `memoryview` object from another Python object that
    /// implements the buffer protocol.
    pub fn from(src: &PyAny) -> PyResult<&PyMemoryView> {
        unsafe {
            src.py()
                .from_owned_ptr_or_err(ffi::PyMemoryView_FromObject(src.as_ptr()))
        }
    }
}
