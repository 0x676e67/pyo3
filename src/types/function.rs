use std::ffi::{CStr, CString};

use crate::derive_utils::PyFunctionArguments;
use crate::exceptions::PyValueError;
use crate::prelude::*;
use crate::{class, ffi, AsPyPointer, PyMethodType};

/// Represents a builtin Python function object.
#[repr(transparent)]
pub struct PyCFunction(PyAny);

pyobject_native_var_type!(PyCFunction, ffi::PyCFunction_Type, ffi::PyCFunction_Check);

impl PyCFunction {
    /// Create a new built-in function with keywords.
    ///
    /// See [raw_pycfunction] for documentation on how to get the `fun` argument.
    pub fn new_with_keywords<'a>(
        fun: ffi::PyCFunctionWithKeywords,
        name: &str,
        doc: &'static str,
        py_or_module: PyFunctionArguments<'a>,
    ) -> PyResult<&'a PyCFunction> {
        let fun = PyMethodType::PyCFunctionWithKeywords(fun);
        Self::new_(fun, name, doc, py_or_module)
    }

    /// Create a new built-in function without keywords.
    pub fn new<'a>(
        fun: ffi::PyCFunction,
        name: &str,
        doc: &'static str,
        py_or_module: PyFunctionArguments<'a>,
    ) -> PyResult<&'a PyCFunction> {
        let fun = PyMethodType::PyCFunction(fun);
        Self::new_(fun, name, doc, py_or_module)
    }

    fn new_<'a>(
        fun: class::PyMethodType,
        name: &str,
        doc: &'static str,
        py_or_module: PyFunctionArguments<'a>,
    ) -> PyResult<&'a PyCFunction> {
        let (py, module) = py_or_module.into_py_and_maybe_module();
        let doc: &'static CStr = CStr::from_bytes_with_nul(doc.as_bytes())
            .map_err(|_| PyValueError::new_err("docstring must end with NULL byte."))?;
        let name = CString::new(name.as_bytes()).map_err(|_| {
            PyValueError::new_err("Function name cannot contain contain NULL byte.")
        })?;
        let def = match fun {
            PyMethodType::PyCFunction(fun) => ffi::PyMethodDef {
                ml_name: name.into_raw() as _,
                ml_meth: Some(fun),
                ml_flags: ffi::METH_VARARGS,
                ml_doc: doc.as_ptr() as _,
            },
            PyMethodType::PyCFunctionWithKeywords(fun) => ffi::PyMethodDef {
                ml_name: name.into_raw() as _,
                ml_meth: Some(unsafe { std::mem::transmute(fun) }),
                ml_flags: ffi::METH_VARARGS | ffi::METH_KEYWORDS,
                ml_doc: doc.as_ptr() as _,
            },
            _ => {
                return Err(PyValueError::new_err(
                    "Only PyCFunction and PyCFunctionWithKeywords are valid.",
                ))
            }
        };
        let (mod_ptr, module_name) = if let Some(m) = module {
            let mod_ptr = m.as_ptr();
            let name = m.name()?.into_py(py);
            (mod_ptr, name.as_ptr())
        } else {
            (std::ptr::null_mut(), std::ptr::null_mut())
        };

        unsafe {
            py.from_owned_ptr_or_err::<PyCFunction>(ffi::PyCFunction_NewEx(
                Box::into_raw(Box::new(def)),
                mod_ptr,
                module_name,
            ))
        }
    }
}

/// Represents a Python function object.
#[repr(transparent)]
pub struct PyFunction(PyAny);

pyobject_native_var_type!(PyFunction, ffi::PyFunction_Type, ffi::PyFunction_Check);
