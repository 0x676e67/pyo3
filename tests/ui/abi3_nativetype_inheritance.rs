//! With abi3, we cannot inherite native types.
use pyo3::prelude::*;
use pyo3::types::PyDict;

#[pyclass(extends=PyDict)]
struct TestClass {}

fn main() {}
