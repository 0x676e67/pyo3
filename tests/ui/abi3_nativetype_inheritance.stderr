error[E0277]: the trait bound `pyo3::ffi::PyDictObject: pyo3::type_object::PySizedLayout<pyo3::types::PyDict>` is not satisfied
  --> $DIR/abi3_nativetype_inheritance.rs:5:1
   |
5  | #[pyclass(extends=PyDict)]
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `pyo3::type_object::PySizedLayout<pyo3::types::PyDict>` is not implemented for `pyo3::ffi::PyDictObject`
   |
  ::: $WORKSPACE/src/type_object.rs:96:22
   |
96 |     type BaseLayout: PySizedLayout<Self::BaseType>;
   |                      ----------------------------- required by this bound in `pyo3::PyTypeInfo`
   |
   = note: required because of the requirements on the impl of `pyo3::type_object::PySizedLayout<pyo3::types::PyDict>` for `pyo3::pycell::PyCellBase<pyo3::types::PyDict>`
   = note: this error originates in an attribute macro (in Nightly builds, run with -Z macro-backtrace for more info)
